File: testing/tests_js/test_037_html_element_frontend.py
"""Node-backed checks for static HTML element failures."""


# @pair form-html:error-reporting
def test_html_element_reports_request_failure_without_masking_original(run_node):
    run_node(
        r'''
const fs = require("node:fs");
const vm = require("node:vm");

(async () => {
class BaseElement {
  constructor(renderer, schema, submission) {
    this.renderer = renderer;
    this.schema = schema;
    this.submission = submission;
  }
}

const original = new Error("HTML request failed");
const captured = [];
const dependencies = {
  BaseElement,
  captureError(...args) {
    captured.push(args);
  },
  ENDPOINTS: {
    html(key, id) {
      return { getContent: `/assets/${key}/html/${id}` };
    },
  },
  request: {
    get() {
      return Promise.reject(original);
    },
  },
};
const context = { dependencies };
vm.createContext(context);

let source = fs.readFileSync("src/script/elements/html.mjs", "utf8");
source = source.replace(
  /import \{([^}]+)\} from "\.\.\/shared";/,
  "const {$1} = dependencies;",
);
source = source.replace(
  /import \{([^}]+)\} from "\.\/base\/baseElement";/,
  "const {$1} = dependencies;",
);
source = source.replace(
  "export class HtmlElement",
  "globalThis.HtmlElement = class HtmlElement",
);
vm.runInContext(source, context);

const target = {};
const element = new context.HtmlElement(
  { form: { key: "form-1", target } },
  { id: "instructions" },
  null,
);
const result = await element._getHtml();

if (result !== "") {
  throw new Error(`Failed HTML request unexpectedly returned ${result}`);
}
if (captured.length !== 1 || captured[0][0] !== original) {
  throw new Error("The original request failure was not reported");
}
if (captured[0][1] !== target) {
  throw new Error("The form target was not attached to the error report");
}
if (captured[0][2]?.schema !== element.schema) {
  throw new Error("The field schema was not attached to the error report");
}
})().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
''',
    )

File: testing/tests_js/test_037_html_element_frontend.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import esmock from "esmock";
import { ENDPOINTS } from "../../src/script/shared/endpoints.mjs";
import { createBrowser } from "../utility/js/environment.mjs";

/** @pair form-html:error-reporting */
test("test_html_element_reports_request_failure_without_masking_original", async (t) => {
	createBrowser(t);
	const original = new Error("HTML request failed");
	const captured = [];
	const captureError = t.mock.fn((...args) => captured.push(args));
	const get = t.mock.fn(async () => {
		throw original;
	});
	const { HtmlElement } = await esmock.strict(
		"../../src/script/elements/html.mjs",
		{
			"../../src/script/shared/index.mjs": {
				captureError,
				ENDPOINTS,
				request: { get },
			},
		},
	);
	const target = document.createElement("div");
	const element = new HtmlElement(
		{ form: { key: "form-1", target } },
		{ id: "instructions" },
		null,
	);

	assert.equal(
		await element._getHtml(),
		"",
		"Failed HTML request returned content",
	);
	assert.equal(get.mock.callCount(), 1);
	assert.equal(
		get.mock.calls[0].arguments[0],
		"/assets/form-1/html/instructions",
	);
	assert.equal(captured.length, 1, "Request failure was not reported once");
	assert.equal(captured[0][0], original, "Original request failure was masked");
	assert.equal(
		captured[0][1],
		target,
		"Form target was not attached to the error report",
	);
	assert.equal(
		captured[0][2]?.schema,
		element.schema,
		"Field schema was not attached to the error report",
	);
});
